src/lib/transition_router/transition_router.ml

```diff
@@ -276,20 +276,23 @@ let wait_till_genesis ~logger ~time_controller =
           , `Int (Int64.to_int_exn (Time.Span.to_ms time_till_genesis)) ) ]
       "Node started before genesis: waiting $time_till_genesis milliseconds \
        before running transition router" ;
-    let seconds_to_wait = 30 in
-    let milliseconds_to_wait = Int64.of_int_exn (seconds_to_wait * 1000) in
-    let rec wait_loop tm =
-      if Int64.(tm <= zero) then ()
-      else (
-        Core.Unix.sleep seconds_to_wait ;
-        let tm_remaining = Int64.(tm - milliseconds_to_wait) in
-        Logger.warn logger ~module_:__MODULE__ ~location:__LOC__
-          "Still waiting $tm_remaining milliseconds before running transition \
-           router"
-          ~metadata:[("tm_remaining", `Int (Int64.to_int_exn tm_remaining))] ;
-        wait_loop tm_remaining )
+    let rec logger_loop () =
+      let%bind () = after (Time_ns.Span.of_sec 30.) in
+      let tm_remaining =
+        Time.diff Consensus.Constants.genesis_state_timestamp
+          (Time.now time_controller)
+      in
+      Logger.warn logger ~module_:__MODULE__ ~location:__LOC__
+        "Still waiting $tm_remaining milliseconds before running transition \
+         router"
+        ~metadata:
+          [ ( "tm_remaining"
+            , `Int (Int64.to_int_exn @@ Time.Span.to_ms tm_remaining) ) ] ;
+      logger_loop ()
```

Comment:
This loop still needs to terminate if there is no time remaining, otherwise the logging will continue after the timeout resolves.